src/ansi.rs

```diff
@@ -0,0 +1,1912 @@
+//! ANSI Terminal Stream Parsing.
+
+use alloc::borrow::ToOwned;
+use alloc::string::{String, ToString};
+use alloc::vec::Vec;
+
+use core::convert::TryFrom;
+use core::fmt::Write;
+use core::time::Duration;
+use core::{iter, str};
+
+use log::{debug, trace};
+
+use crate::{Params, ParamsIter};
+
+pub trait SyncHandler: Default {
+    /// Expiration time of the synchronized update.
+    fn update_timeout(&mut self, _: Option<Duration>);
+    fn pending_timeout(&self) -> bool;
+}
+
+#[cfg(not(feature = "no_std"))]
+pub type DefaultSyncHandler = StdTimeProvider;
+
+#[cfg(not(feature = "no_std"))]
+#[derive(Default)]
+pub struct StdTimeProvider {
+    timeout: Option<std::time::Instant>,
+}
+
+#[cfg(not(feature = "no_std"))]
+impl SyncHandler for StdTimeProvider {
+    fn update_timeout(&mut self, duration: Option<Duration>) {
+        use std::time::Instant;
+        self.timeout = duration.map(|e| Instant::now() + e);
+    }
+
+    fn pending_timeout(&self) -> bool {
+        self.timeout.is_some()
+    }
+}
+
+#[cfg(feature = "no_std")]
+pub type DefaultSyncHandler = NullTimeProvider;
+
+/// Time provider which always returns 0.
+#[derive(Default)]
+pub struct NullTimeProvider;
+
+impl SyncHandler for NullTimeProvider {
+    fn update_timeout(&mut self, _: Option<Duration>) {}
+
+    fn pending_timeout(&self) -> bool {
+        false
+    }
+}
```

Comment:
What exactly is this supposed to be? Are you suggesting to just break everything on no_std without users being able to know what's going on?

Reply:
Alternative would be to not have a default sync handler for `no_std`, which should be fine and would force the user to create their own handler.

Reply:
And how would you suggest a `no_std` implementation to look like?

Reply:
It would be nice to have an alacritty PR along the way.

Reply:
> And how would you suggest a `no_std` implementation to look like?

With `no_std` it really depends on where you intend to use the crate. For example, if you intend of using the crate in a kernel, then the implementation would differ from a case where you are using the `libc` to get the timer functionality. 

In any case, it will look very similar to the default implementation of the sync handler that is currently provided for `std`. Here is roughly what the sync handler would look like by using the `libc`:
```rust
struct MySyncHandler {
     timeout: Option<libc::timespec>,
}

impl vte::SyncHandler for MySyncHandler {
    fn update_timeout(&mut self, duration: Option<Duration>) {
         let mut timespec = core::mem::zeroed();
         let now = cvt(unsafe { libc::clock_gettime(CLOCK_MONOTONIC, &mut timespec) }).unwrap();

         self.timeout = duration.map(|e| timespec);
    }

    #[inline]
    fn pending_timeout(&self) -> bool {
         self.timeout.is_some()
    }
}
```

Reply:
I modified the `Processor` type in https://github.com/alacritty/vte/pull/85/commits/c70bc9645553c8c2238f411327e8532ff83234b2 to require a sync handler if it is being used in a `no_std` environment. Otherwise, the default `StdSyncHandler` is used. This change should alleviate the worry of unexpected issues arising in `no_std` environments without users being able to identify the cause.